File: 2021/day1.ts
import { List } from '../common'

export default (rows: string[]) => {
  const input = rows.map(Number)

  const result1 = List.zipPairs(input).filter(([a, b]) => b > a).length
  const result2 = List.zip(input, input.slice(3)).filter(([a, b]) => b > a).length

  return [result1, result2, 1655, 1683]
}

File: 2024/day1.fs
module Year2024.Day1

let countOf value =
    Map.tryFind value >> Option.defaultValue 0

let solve = DayUtils.solveDay (fun input ->
    let numbers = input |> Array.map Input.parseAllInts |> array2D

    let left = numbers[*, 0] |> Array.sort
    let right = numbers[*, 1] |> Array.sort

    let rightCounts = right |> Array.countBy id |> Map

    let result1 = Array.zip left right |> Array.sumBy (fun (l, r) -> abs (r - l))
    let result2 = left |> Array.sumBy (fun l -> l * countOf l rightCounts)

    result1, result2, 2742123, 21328497)
